### PR TITLE
Add test data cache

### LINKED ISSUE #1761: Unit tests are very slow to run
## Expected Behaviour
Tests should run quickly this allows developers to run them often as they work

## Actual Behaviour
Due to the fact it installs the framework multiple times during the tests, they are very slow to run.

## Versions
  - Authoring Tool Version: 0.4.0

### PR DESCRIPTION
**Pull request by taylortom**
Fixes #1761

### CODE REVIEW

**Comment by canstudios-louisem** [comment]
In the interest of keeping 0.4.1 light and easier to test/get out I think we should leave this for 0.5.0.

**Comment by taylortom** [comment]
Need to investigate how much the GitHub API rate limit affects the running of the tests

File: lib/installHelpers.js [hunk 1]
```diff
@@ -295,13 +294,13 @@ function installFramework(opts, callback) {
...
-  if(!fs.existsSync(opts.directory) || opts.force) {
-    return async.applyEachSeries([
-      cloneRepo,
-      updateFramework
-    ], opts, callback);
+  if(fs.existsSync(opts.directory) && !opts.force) {
+    return updateFramework(opts, callback);
   }
-  updateFramework(opts, callback);
+  return async.applyEachSeries([
```
Comment (tomgreenfield) [inline lib/installHelpers.js]: Is the `return` necessary here?

Comment (taylortom) [reply lib/installHelpers.js]: No. I quite like it there though.

File: test/entry.js [hunk 2]
```diff
@@ -14,43 +15,43 @@ var testData = require('./testData.json');
...
-
-          createTestUser(testData.testUser, function(error, user) {
-            if(error) {
-              return cb(error);
-            }
-            testData.testUser._id = user._id;
-            app.rolemanager.assignRoleByName('Super Admin', user._id, cb);
-          });
-        });
+        cb();
```
Comment (tomgreenfield) [inline test/entry.js]: Could be `app.on('serverStarted', cb);`.

File: lib/installHelpers.js [hunk 3]
```diff
@@ -215,10 +215,8 @@ function checkLatestAdaptRepoVersion(repoName, versionLimit, callback) {
         error = 'GitubAPI did not respond with a 200 status code.';
       }
     }
-    // exit, but just log the error
     if (error) {
-      log(`Couldn't check latest version of ${repoName}\n${error}`);
-      return callback();
+      return callback(`Couldn't check latest version of ${repoName}. ${error}`);
```
[the remark this answers is not in the document]
Comment (taylortom) [reply lib/installHelpers.js]: Have replaced all error strings with Error instances

File: test/entry.js [hunk 4]
```diff
@@ -121,6 +122,35 @@ function removeTestData(done) {
...
+  app.configuration.setConfig('masterTenantName', testData.testTenant.name);
+
+  var src = path.join(TEST_CACHE_DIR, Folders.Framework);
+
+  function _copyFramework() {
+    var dest = path.join(app.configuration.serverRoot, Folders.Temp, testData.testTenant._id);
+    async.series([
+      async.apply(fs.remove, dest),
+      async.apply(fs.ensureDir, dest),
+      async.apply(fs.symlink, src, path.join(dest, Folders.Framework), 'dir')
```
Comment (lc-thomasberger) [inline test/entry.js]: this throws a EPERM Error on Windows:
Error: EPERM: operation not permitted, symlink 'D:\AT-Project\at_repo\adapt_authoring\test\.testcache\adapt_framework' -> 'D:\AT-Project\at_repo\adapt_authoring\temp\000000000000000000000000\adapt_framework'

Creating a junction fixes the issue on my machine:
async.apply(fs.symlink, src, path.join(dest, Folders.Framework), 'junction')

File: test/entry.js [hunk 5]
```diff
@@ -14,43 +15,43 @@ var testData = require('./testData.json');
 var testConfig = require('./testConfig.json');
 
 var app = origin();
+var Folders = require('../lib/outputmanager').Constants.Folders;
 
+var TEST_CACHE_DIR = path.join(__dirname, '.testcache');
```
Comment (lc-thomasberger) [inline test/entry.js]: should this directory be added to .gitignore?

File: test/entry.js (hunk 4 above)
Comment (taylortom) [reply test/entry.js]: Have changed this, let me know if it's working now :)

**Pull request merged**